domain/relation/service/relation.go

```diff
@@ -501,30 +507,40 @@ func (s *Service) EnterScope(
 	}
 
 	// Enter the unit into the relation scope.
-	err := s.st.EnterScope(ctx, relationUUID, unitName, settings)
+	relUnitUUID, err := s.st.EnterScope(ctx, relationUUID, unitName, settings)
 	if errors.Is(err, relationerrors.RelationUnitAlreadyExists) {
 		return nil
 	} else if err != nil {
 		return errors.Capture(err)
 	}
 
+	defer func() {
+		if err != nil {
+			// If there is any failure attempting to create the subordinate unit,
+			// roll back the relation unit creation. EnterScope may only succeed
+			// if the subordinate unit creation succeeds when the subordinate unit
+			// should be created.
+			backoutErr := s.st.DeleteRelationUnit(ctx, relUnitUUID)
+			if backoutErr != nil {
+				s.logger.Errorf(ctx, "rollback of relation unit on failed subordinate creation: %w", backoutErr)
+			}
+		}
+	}()
+
 	// Check if a subordinate unit needs creating.
 	subID, err := s.st.NeedsSubordinateUnit(ctx, relationUUID, unitName)
 	if err != nil {
 		return errors.Capture(err)
 	} else if subID != nil {
 		// Create the required unit on the related subordinate application.
-		//
-		// TODO(aflynn): In 3.6 the subordinate was created in the same
-		// transaction as the principal entering scope. This is not the case
-		// here. If the subordinate creation fails, there should be some retry
-		// mechanism, or a rollback of enter scope.
 		if subordinateCreator == nil {
-			return errors.Errorf("subordinate creator is nil")
+			err = errors.Errorf("subordinate creator is nil")
+			return errors.Capture(err)
 		}
-		err := subordinateCreator.CreateSubordinate(ctx, *subID, unitName)
-		if err != nil {
-			return errors.Errorf("creating subordinate unit on application %q: %w", *subID, err)
+
+		if err = subordinateCreator.CreateSubordinate(ctx, *subID, unitName); err != nil {
```

Comment:
In the future, this should be either driven by a worker, or the relation join that needs a subordinate to be handled by the application domain with arguments created by the relation domain.